### ISSUE TITLE
Returning multiple dataframes over a large file where the entire data does not fit in memory

### ISSUE THREAD

**Issue by cyr** [pola-rs/polars#3260]
Really promising project, very impressive overall!

I have an issue that I don't seem to be able to handle in any sane way. I have a fairly large (multi-gb) csv file. In a simple case I just want to convert it to parquet, but since it does not fit in memory, this fails.

What would be helpful is if there was a way to return a dataframe over a slice of data, and then call for another dataframe once whatever processing (or not) is done on that, until the whole file is streamed through. 

As is, I would have to do separate calls to create a dataframe with a slice and manually move it forward. This seems to take longer and longer the deeper into the file it looks - which is to be expected of course, but is not ideal.

I'm imagining something like this:
````python
for df in pl.scan_csv("somefile.csv").iterative_slice(10000):
  #df.do_stuff()
```` 

Sorry if this is already possible in some way - and if there is I would like to know! :)

**Comment by ghuls** [comment]
You might need to use pyarrows streaming CSV reader for now: https://arrow.apache.org/docs/python/generated/pyarrow.csv.CSVStreamingReader.html

https://stackoverflow.com/questions/68555085/how-can-i-chunk-through-a-csv-using-arrow

You can convert pyarrow table to a polars dataframe or write the pyarrow table to parquet directly from pyarrow.

**Comment by cyr** [comment]
That is definitely doable. The downside would be that any defined columns and transforms would need to applied per chunk, correct? I suppose it doesn't matter a whole lot.

Is there a way to write to parquet in a rowgroup by rowgroup way? That is, each chunk ending up in a rowgroup instead of as separate files.

**Comment by ghuls** [comment]
Yes, if you don't have enough memory to load the whole dataset at once, you will need to process them by chunk.

You can write row groups with pyarrow when writing to parquet and read them by row groups again.

https://arrow.apache.org/docs/python/parquet.html#finer-grained-reading-and-writing

**Comment by ghuls** [comment]
https://www.orchest.io/blog/the-great-python-dataframe-showdown-part-1-demystifying-apache-arrow givees a solution for your problem:

> A possible use case is converting a huge CSV file to Parquet by batches, as follows:

```python
import pyarrow as pa
import pyarrow.parquet as pq

from pyarrow.csv import open_csv

mmap = pa.memory_map("/data/yellow_tripdata_2015-01.csv")

# Create a CSVStreamingReader from the memory-mapped file
reader = open_csv(mmap)
‍
# "Rewind" the CSV file
mmap.seek(0)
reader = open_csv(mmap)
‍
# Open parquet file for writing with same schema as the CSV file
with pq.ParquetWriter("/data/yellow_tripdata_2015-01.parquet", reader.schema) as writer:
    while True:
        try:
            batch = reader.read_next_batch()
            writer.write_batch(batch)
        except StopIteration:
            break

# Load data directly from Parquet
reloaded_nyc = pq.read_table("/data/yellow_tripdata_2015-01.parquet")
```

**Comment by cyr** [comment]
This does work quite well, thank you! 

Is it possible to apply transforms via polars and still do it this way though? The underlying pyarrow table does not pick up changes made. I could probably create the schema on the fly, but it would require some fiddling around. And this would not take into account things such as type conversions or adding on new columns, etc.

I think it would be really helpful to have some way to iteratively produce tables like pyarrow is doing here, but from within polars.

**Comment by cyr** [comment]
I'll add that I was mistaken, using to_arrow() does allow me to do all the things required, after doing transformations. I still think it's not a bad idea to do this, but as this works, I won't die if you decide it's not worth implementing. Close this if you want :)

**Comment by daviewales** [comment]
I was hoping to be able to do `pl.scan_csv('big.csv').write_parquet('big.parquet')`, and have Polars automatically handle the chunking, but I'll look into the Arrow method for now.

**Comment by alexander-beedie** [comment]
> I was hoping to be able to do `pl.scan_csv('big.csv').write_parquet('big.parquet')`, and have Polars automatically handle the chunking, but I'll look into the Arrow method for now.

This can now be handled by the [sink_parquet](https://pola-rs.github.io/polars/py-polars/html/reference/lazyframe/api/polars.LazyFrame.sink_parquet.html) feature, eg:

```python
ldf = pl.scan_csv("/path/to/my_larger_than_ram_file.csv")  
ldf.sink_parquet("out.parquet")  
```
